File: MiniDOMBase/DMD/Sistema/Backup/Catalog/CBackupItem.vb
Imports Ionic.Zip

Partial Class Sistema

    Public Class CBackupItem
        Implements XML.IDMDXMLSerializable

        Public m_Owner As CBackup
        Public PercorsoOrigine As String
        Public FileDestinazione As String
        Public DimensioneOrigine As Long
        Public DimensioneCompressa As Long
        Public DataUltimaModifica As Date
        Public DataCompressione As Date
        Public Messages As String

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
        End Sub

        Public ReadOnly Property Owner As CBackup
            Get
                Return Me.m_Owner
            End Get
        End Property

        Protected Friend Overridable Sub SetOwner(ByVal owner As CBackup)
            Me.m_Owner = owner
        End Sub

        
        Public Sub Comprimi(ByVal percorsoOrigine As String, ByVal percorsoBase As String)
            Me.PercorsoOrigine = Sistema.FileSystem.GetRelativePath(percorsoOrigine, percorsoBase)
            Me.DimensioneOrigine = Sistema.FileSystem.GetFileSize(percorsoOrigine)
            Me.DataUltimaModifica = Sistema.FileSystem.GetLastModifiedTime(percorsoOrigine)

            Dim out As New System.IO.StringWriter
            Dim zip As New ZipFile
            zip.StatusMessageTextWriter = out
            zip.ZipErrorAction = ZipErrorAction.Skip
            'zip.AddDirectory(Sistema.ApplicationContext.WorkingFolder)
            zip.AddFile(percorsoOrigine, "")
            zip.CompressionLevel = Me.m_Owner.CompressionLevel
            zip.CompressionMethod = Me.m_Owner.CompressionMethod
            zip.UseZip64WhenSaving = Zip64Option.AsNecessary

            Dim dir As String = Sistema.FileSystem.CombinePath(Sistema.FileSystem.GetFolderName(Me.m_Owner.FileName), Sistema.FileSystem.GetBaseName(Me.m_Owner.FileName))
            Dim stream As System.IO.Stream = Nothing
            SyncLock Me.m_Owner
                Me.FileDestinazione = ASPSecurity.GetRandomKey(8) & ".ZIP"
                While Sistema.FileSystem.FileExists(Sistema.FileSystem.CombinePath(dir, Me.FileDestinazione))
                    Me.FileDestinazione = ASPSecurity.GetRandomKey(8) & ".ZIP"
                End While
                stream = New System.IO.FileStream(Sistema.FileSystem.CombinePath(dir, Me.FileDestinazione), System.IO.FileMode.Create)
            End SyncLock
            zip.Save(stream)
            Me.DataCompressione = DateUtils.Now
            Me.DimensioneCompressa = Sistema.FileSystem.GetFileSize(Sistema.FileSystem.CombinePath(dir, Me.FileDestinazione))
            stream.Dispose()
            Me.Messages = out.ToString
            out.Dispose()
        End Sub

        Public Sub Decomprimi(ByVal percorsoBase As String)
            Dim dir As String = Sistema.FileSystem.CombinePath(Sistema.FileSystem.GetFolderName(Me.m_Owner.FileName), Sistema.FileSystem.GetBaseName(Me.m_Owner.FileName))
            Dim f As String = Sistema.FileSystem.GetFileName(Me.FileDestinazione)
            Dim zip As New ZipFile(Sistema.FileSystem.CombinePath(dir, f))
            'AddHandler zip.ExtractExceptio, AddressOf zipExtractError
            Dim p As String = Sistema.FileSystem.GetAbsolutePath(Me.PercorsoOrigine, percorsoBase)
            p = Sistema.FileSystem.GetFolderName(p)
            Sistema.FileSystem.CreateRecursiveFolder(p)
            zip.ExtractAll(p, ExtractExistingFileAction.OverwriteSilently)
            zip.Dispose()
        End Sub

        'Public Sub Delete()
        '    Dim dir As String = Sistema.FileSystem.CombinePath(Sistema.FileSystem.GetFolderName(Me.m_Owner.FileName), Sistema.FileSystem.GetBaseName(Me.m_Owner.FileName))
        '    Dim f As String = Sistema.FileSystem.GetFileName(Me.FileDestinazione)
        '    Dim p As String = Sistema.FileSystem.GetAbsolutePath(Me.PercorsoOrigine, percorsoBase)
        '    p = Sistema.FileSystem.GetFolderName(p)
        '    Sistema.FileSystem.CreateRecursiveFolder(p)
        '    zip.ExtractAll(p, ExtractExistingFileAction.OverwriteSilently)
        '    zip.Dispose()
        'End Sub


        Public Sub SetFieldInternal(fieldName As String, fieldValue As Object) Implements XML.IDMDXMLSerializable.SetFieldInternal
            Select Case fieldName
                Case "PercorsoOrigine" : Me.PercorsoOrigine = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "FileDestinazione" : Me.FileDestinazione = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "DimensioneOrigine" : Me.DimensioneOrigine = XML.Utils.Serializer.DeserializeLong(fieldValue)
                Case "DimensioneCompressa" : Me.DimensioneCompressa = XML.Utils.Serializer.DeserializeLong(fieldValue)
                Case "DataUltimaModifica" : Me.DataUltimaModifica = XML.Utils.Serializer.DeserializeDate(fieldValue)
                Case "DataCompressione" : Me.DataCompressione = XML.Utils.Serializer.DeserializeDate(fieldValue)
                Case "Messages" : Me.Messages = XML.Utils.Serializer.DeserializeString(fieldValue)
            End Select
        End Sub

        Public Sub XMLSerialize(writer As XML.XMLWriter) Implements XML.IDMDXMLSerializable.XMLSerialize
            writer.WriteAttribute("PercorsoOrigine", Me.PercorsoOrigine)
            writer.WriteAttribute("FileDestinazione", Me.FileDestinazione)
            writer.WriteAttribute("DimensioneOrigine", Me.DimensioneOrigine)
            writer.WriteAttribute("DimensioneCompressa", Me.DimensioneCompressa)
            writer.WriteAttribute("DataUltimaModifica", Me.DataUltimaModifica)
            writer.WriteAttribute("DataCompressione", Me.DataCompressione)
            writer.WriteTag("Messages", Me.Messages)
        End Sub

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub
    End Class

End Class

File: MiniDOMBase/DMD/Sistema/Backup/Catalog/CBackupItem.cs
using System;
using System.Collections;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using static minidom.Sistema;
using static minidom.Anagrafica;
using minidom.repositories;
using System.Collections.Generic;
using DMD.Zip;

namespace minidom
{
    public partial class Sistema
    {

        /// <summary>
        /// Elemento nel catalogo del backup
        /// </summary>
        [Serializable]
        public class CBackupItem
            : IDMDXMLSerializable
        {

            /// <summary>
            /// Possessore
            /// </summary>
            [NonSerialized]  public CBackup m_Owner;

            /// <summary>
            /// Percorso completo del file originale
            /// </summary>
            public string PercorsoOrigine;

            /// <summary>
            /// Percorso completo del file destinazione
            /// </summary>
            public string FileDestinazione;

            /// <summary>
            /// Dimensione in bytes del file originale
            /// </summary>
            public long DimensioneOrigine;

            /// <summary>
            /// Dimensione in bytes del file compresso
            /// </summary>
            public long DimensioneCompressa;

            /// <summary>
            /// Data dell'ultima modifica del file originale
            /// </summary>
            public DateTime DataUltimaModifica;

            /// <summary>
            /// Data di compressione
            /// </summary>
            public DateTime DataCompressione;

            /// <summary>
            /// Messaggi generati dal compressore per l'elemento
            /// </summary>
            public string Messages;

            /// <summary>
            /// Costruttore
            /// </summary>
            public CBackupItem()
            {
                //DMDObject.IncreaseCounter(this);
            }

            /// <summary>
            /// Riferimento al backup
            /// </summary>
            public CBackup Owner
            {
                get
                {
                    return m_Owner;
                }
            }

            /// <summary>
            /// Imposta il possessore
            /// </summary>
            /// <param name="owner"></param>
            protected internal virtual void SetOwner(CBackup owner)
            {
                m_Owner = owner;
            }

            /// <summary>
            /// Comprime l'elemento singolarmente
            /// </summary>
            /// <param name="percorsoOrigine"></param>
            /// <param name="percorsoBase"></param>
            public void Comprimi(string percorsoOrigine, string percorsoBase)
            {
                this.PercorsoOrigine = FileSystem.GetRelativePath(percorsoOrigine, percorsoBase);
                this.DimensioneOrigine = FileSystem.GetFileSize(percorsoOrigine);
                this.DataUltimaModifica = FileSystem.GetLastModifiedTime(percorsoOrigine);
                using (var @out = new System.IO.StringWriter())
                {
                    using (var zip = new ZipFile())
                    {
                        zip.StatusMessageTextWriter = @out;
                        zip.ZipErrorAction = ZipErrorAction.Skip;
                        // zip.AddDirectory(Sistema.ApplicationContext.WorkingFolder)
                        zip.AddFile(percorsoOrigine, "");
                        zip.CompressionLevel = (CompressionLevel)m_Owner.CompressionLevel;
                        zip.CompressionMethod = (CompressionMethod)m_Owner.CompressionMethod;
                        zip.UseZip64WhenSaving = Zip64Option.AsNecessary;
                        string dir = FileSystem.CombinePath(FileSystem.GetFolderName(m_Owner.FileName), FileSystem.GetBaseName(m_Owner.FileName));
                        lock (this.m_Owner)
                        {
                            this.FileDestinazione = DMD.Strings.GetRandomString(8) + ".ZIP";
                            while (FileSystem.FileExists(FileSystem.CombinePath(dir, FileDestinazione)))
                                FileDestinazione = DMD.Strings.GetRandomString(8) + ".ZIP";
                            using (var stream = new System.IO.FileStream(FileSystem.CombinePath(dir, FileDestinazione), System.IO.FileMode.Create))
                            {
                                zip.Save(stream);
                            }
                        }
                        DataCompressione = DMD.DateUtils.Now();
                        DimensioneCompressa = FileSystem.GetFileSize(FileSystem.CombinePath(dir, FileDestinazione));

                        this.Messages = @out.ToString();
                    }
                }
            }

            /// <summary>
            /// Decomprime l'elemento
            /// </summary>
            /// <param name="percorsoBase"></param>
            public void Decomprimi(string percorsoBase)
            {
                string dir = FileSystem.CombinePath(FileSystem.GetFolderName(m_Owner.FileName), FileSystem.GetBaseName(m_Owner.FileName));
                string f = FileSystem.GetFileName(FileDestinazione);
                using (var zip = new ZipFile(FileSystem.CombinePath(dir, f)))
                {
                    // AddHandler zip.ExtractExceptio, AddressOf zipExtractError
                    string p = FileSystem.GetAbsolutePath(PercorsoOrigine, percorsoBase);
                    p = FileSystem.GetFolderName(p);
                    FileSystem.CreateRecursiveFolder(p);
                    zip.ExtractAll(p, ExtractExistingFileAction.OverwriteSilently);
                }
            }

            // Public Sub Delete()
            // Dim dir As String = Sistema.FileSystem.CombinePath(Sistema.FileSystem.GetFolderName(Me.m_Owner.FileName), Sistema.FileSystem.GetBaseName(Me.m_Owner.FileName))
            // Dim f As String = Sistema.FileSystem.GetFileName(Me.FileDestinazione)
            // Dim p As String = Sistema.FileSystem.GetAbsolutePath(Me.PercorsoOrigine, percorsoBase)
            // p = Sistema.FileSystem.GetFolderName(p)
            // Sistema.FileSystem.CreateRecursiveFolder(p)
            // zip.ExtractAll(p, ExtractExistingFileAction.OverwriteSilently)
            // zip.Dispose()
            // End Sub

            /// <summary>
            /// Deserializzazione xml
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            public void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "PercorsoOrigine":
                        {
                            PercorsoOrigine = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "FileDestinazione":
                        {
                            FileDestinazione = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "DimensioneOrigine":
                        {
                            DimensioneOrigine = (long)DMD.XML.Utils.Serializer.DeserializeLong(DMD.Strings.CStr(fieldValue));
                            break;
                        }

                    case "DimensioneCompressa":
                        {
                            DimensioneCompressa = (long)DMD.XML.Utils.Serializer.DeserializeLong(DMD.Strings.CStr(fieldValue));
                            break;
                        }

                    case "DataUltimaModifica":
                        {
                            DataUltimaModifica = (DateTime)DMD.XML.Utils.Serializer.DeserializeDate(fieldValue);
                            break;
                        }

                    case "DataCompressione":
                        {
                            DataCompressione = (DateTime)DMD.XML.Utils.Serializer.DeserializeDate(fieldValue);
                            break;
                        }

                    case "Messages":
                        {
                            Messages = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }
                }
            }

            /// <summary>
            /// Serializzazione xml
            /// </summary>
            /// <param name="writer"></param>
            public void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("PercorsoOrigine", PercorsoOrigine);
                writer.WriteAttribute("FileDestinazione", FileDestinazione);
                writer.WriteAttribute("DimensioneOrigine", DimensioneOrigine);
                writer.WriteAttribute("DimensioneCompressa", DimensioneCompressa);
                writer.WriteAttribute("DataUltimaModifica", DataUltimaModifica);
                writer.WriteAttribute("DataCompressione", DataCompressione);
                writer.WriteTag("Messages", Messages);
            }

            //~CBackupItem()
            //{
            //    //DMDObject.DecreaseCounter(this);
            //}
        }
    }
}